text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,391 @@
+---
+feature name: cli-framework-compatibility-strategy
+start date: February 2nd, 2020
+rfc pr: https://github.com/aws/aws-cdk-rfcs/pull/111
+related issue: https://github.com/aws/aws-cdk-rfcs/issues/110
+---
+
+<!-- replace the blockquoted sections with your content -->
+# Summary
+
+There are currently two promises we try to keep with regards to framework and cli compatibility:
+
+- CLI versions should always support older framework versions. That is, upgrading the CLI should never require an upgrade of the framework.
+- Framework versions might require an upgrade of the CLI. In which case, we provide instructions as to which CLI version should be used.
+
+This RFC details the necessary strategy and mechanism for ensuring that we keep those promises when releasing new versions.
+
+# Motivation
+
+After implementing this RFC, we can expect the following outcomes:
+
+- An incompatible change in the CLI won't make it to the release.
```

Comment:
Another thing we discussed is that currently the cx-api module is marked "experimental" which means it bypasses our jsii API compatibility tests. If we mark it "stable" (which it has to be by definition), then we will enjoy the compat tests and it will essentially verify that new CLI must always support old frameworks.

Reply:
+2

Reply:
Can you talk me through how that will help? I'm not sure I understand.

The way I understand this to help at all is accidentally, because the structs that make up the file schema *happen to be* exposed. 

It would also stability-check the API of something like `CloudAssemblyBuilder`, which is a completely separate issue from CX compatibility.

Reply:
@rix0rrr Not entirely sure what you mean by accidental.

The cx-api is a module just like any other, that provides an external protocol to be consumed by the CLI. Running jsii compatibility checks on that API, is a way to make sure we don't break  it. 

Since the CLI depends on it, it will also help make sure that new CLI versions are able to consume older cloud assemblies. 

Honestly, i'm still a bit vague on the specifics, but it feels like running compatibility checks on cx-api should make just as much sense as running compatibility checks for all our other modules.

Reply:
I'm not convinced that jsii compatibility checks are necessarily going to help us. If the API has:

```ts
function writeFile(data: string[]): void;
function readFile(): string[];
```

There is no guarantee that the file written by v1 can be read by v2, even though the APIs didn't change.

There is additional state on the disk that cannot be checked by the API compatibility checker.

Reply:
Think of the cloud assembly as the external interface of a CDK app, which is consumed by the CLI. It is defined through a typescript interface and as long as this interface doesn’t change in a breaking way, the CLI should be able to consume older versions safely. 

For example, say we add an attribute to docker assets. If we add this attribute as a required field, and we modify the CLI to consume it, everything will compile, but now the CLI won’t support older versions because it assumes this attribute exists. If we add it as an optional field, the typescript compiler will require that CLI code will only consume this attribute conditionally if it’s defined. This will prompt backwards compatible behavior.

Luckily we have a very awesome tool that can help us enforce that changes to jsii APIs are backwards compatible. All we need is to use it for the cloud assembly api!

Reply:
> There is no guarantee that the file written by v1 can be read by v2, even though the APIs didn't change.

The written file is supposed to be a serialized representation of the API. So as long as the API is backwards compatible, deserialization should work. 
We shouldn't be having any critical state in the file system that isn't described in the protocol, this implies the protocol isn't defined properly, and there is a leak somewhere. 

The cloud-assembly is just a protocol, that has consumers, coincidentally, one of those consumers is our own CLI. You can imagine other consumers such as custom CLI tools or web servers that accept a cloud-assembly and do something with it. 

A protocol shouldn't be aware of its consumers, and shouldn't be making any promises to any specific ones. The only thing a protocol can do, is ensure backwards compatibility with **itself**, which what the JSII check will guarantee. 

Im still gathering my thoughts on this, but it seems like if think about cx-api as an isolated protocol, things become much clearer. This also implies that the cx-api code shouldn't have any references or awareness of the CLI, only the other way around.

Reply:
By the way, the fact our CLI is written in `typescript`, is some what coincidental as well. I mean, design wise, we shouldn't have any limitation for re-writing the CLI in any other language, as long as there is a well defined protocol. In that sense, the fact the cli literally depends on `cx-api` is also somewhat of a leak in my opinion. This relates to what @eladb was talking about with regards to actually having a copy of the protocol, and not depend on it. I am not entirely verse in it, but I think `protobuff` is a good example, where they have the `protoc` compiler that generates code for clients to consume the code in many languages, essentially creates this copy.

Reply:
> In that sense, the fact the cli literally depends on cx-api is also a leak in my opinion.

It is, but I don't think it's a giant problem. The alternatives to this I can see just add busywork for dubious benefit.

I see where you both are coming from with the API compatibility checker == protocol compatibility checker. I agree in principle, just not out-of-hand convinced of  the proposed solution.

YES, we can check protocol compatibility this way (or at least, somewhat). To do that, we need to check the structs that make up the protocol schema.

NO, I'm not sure it's as simple as turning on API compatibility checking for the `@aws-cdk/cx-api` module. 

For one, there's more in that module than just the structs that make up the protocol schema.

For second, compatibility is principally computed for *functions*, not for *data*. The constraints on a data structure follow from how they are used in functions. Are they used as function parameters? Can't make an optional field required! (You may only weaken preconditions). Are they used as return value from functions? Can't make a required field optional! (You may only strengthen postconditions).

The compatibility requirements on data structures are computed based on how they are used in relation to functions, and I'm not sure the functions currently in the `cx-api` module currently take all the schema structs in the right position for the rules that we desire to be applied.

To apply this solution, we would need to slightly change the compatibility checker to be able to check a subset of structs/assuming a weakening/strengthening constraint for them... or move the schema structs (and ONLY the schema structs) to a separate module, so we can compat-check those. `cx-api` then becomes a set of convenience routines that happen to read/write this schema in TypeScript. Or the other way around, `cx-api` becomes only the data schema and the TypeScript routines that operate on them move to a separate package.

-----

BTW, given the new approach of the CLI at least matching the framework, I believe we are only allowed to *strengthen* the schema. We are allowed to add new fields, but we're not allowed to take any away.

That corresponds for the compatibility checker to a function existing that does this:

```ts
function loadAssembly(): Schema;
```

And *NO* function that looks like this:

```ts
function writeAssembly(schema: Schema): void;
```

Reply:
Ok, we might be getting a bit too theoretic on this one? I'm currently working on writing up breakage scenarios and explaining how those scenarios will be detected. 

Im also investigating exactly what does `cx-api` provide and how should it possibly be restructured and why.

Also, I think i need some more knowledge on the checker itself, so i'll be learning a bit about it as well. I do however like the idea of creating a data only package that includes the structs that make up the protocol and nothing more. 

FYI, I am not saying the protocol check is the **only** thing we need in order to ensure CLI isn't broken. We also need standard regressions tests that make sure the CLI actually properly **uses** the protocol. 

In any case, I hope we can have a more concrete discussion in a couple of days. 

Sound good?

Reply:
I really like @rix0rrr’s suggestion to extract the cloud assembly protocol into a separate module and add a single class there with a “loadAssembly” method, and then activate compatibility checks on this new module. This sounds like a very simple way to leverage our existing tools and getting most of the benefits.

Reply:
Yeap, thats what i'll be aiming for.

Reply:
I've been thinking about this some more, and I think the contract we need to validate is actually `saveAssemly(asm)`. That is, the assembly is used as an *input* and not as an *output*. This will actually enforce the right behavior in terms of compatibility checks.

For example, when a struct is used as an output (i.e. the "loadAssembly" option), it is ok for it to add required fields. But in our case, we don't want this to be allowed.

Therefore, I was wondering whether the mental model I proposed of *"the cloud assembly is the output interface of the app"* might be the wrong one given our compatibility model. Maybe the right mental model is ***"the cloud assembly is the input interface of the CLI"***.

Bottom line, I feel strongly that we should probably extract the cloud assembly into it's own module so we can make compatibility claims on it without the noise of the rest of the crap in cx-api (which we should also think about).